File: src/tasks/cli/command/add/AddParser.java
package tasks.cli.command.add;

import static tasks.cli.parser.ParserUtil.extractSuccessfulResults;
import static tasks.cli.parser.ParserUtil.getOptionValues;
import static tasks.cli.parser.ParserUtil.validateParsedTasks;

import java.util.Optional;
import omnia.data.cache.Memoized;
import omnia.data.structure.List;
import omnia.data.structure.immutable.ImmutableList;
import org.apache.commons.cli.CommandLine;
import tasks.cli.parser.ArgumentFormatException;
import tasks.cli.parser.ParserUtil;
import tasks.cli.parser.CommandParser;
import tasks.cli.parser.Parser;
import tasks.cli.parser.ParseResult;
import tasks.model.Task;

/** Command line argument parser for the Add command. */
public final class AddParser implements CommandParser<AddArguments> {
  private final Memoized<? extends Parser<? extends List<? extends ParseResult<? extends Task>>>> taskParser;

  public AddParser(
      Memoized<? extends Parser<? extends List<? extends ParseResult<? extends Task>>>>
          taskParser) {
    this.taskParser = taskParser;
  }

  @Override
  public AddArguments parse(CommandLine commandLine) {
    /*
     * 1st param must be description
     * 2+ params not supported
     * optional blocking tasks
     * optional blocked tasks
     */
    List<String> argsList = ImmutableList.copyOf(commandLine.getArgList());
    String taskDescription = extractTaskDescriptionFrom(argsList)
        .orElseThrow(
            () -> new ArgumentFormatException("Task description not defined"));
    ParserUtil.assertNoExtraArgs(commandLine, AddCommand.COMMAND_PARAMETERS.value());

    List<? extends ParseResult<? extends Task>> afterTasks =
        taskParser.value().parse(getOptionValues(commandLine, AddCommand.AFTER_OPTION.value()));
    List<? extends ParseResult<? extends Task>> beforeTasks =
        taskParser.value().parse(getOptionValues(commandLine, AddCommand.BEFORE_OPTION.value()));

    validateParsedTasks(
        ImmutableList.<ParseResult<?>>builder()
            .addAll(afterTasks)
            .addAll(beforeTasks)
            .build());

    return new AddArguments(
        taskDescription,
        extractSuccessfulResults(afterTasks),
        extractSuccessfulResults(beforeTasks));
  }

  private static Optional<String> extractTaskDescriptionFrom(List<String> args) {
    return args.count() < 1 ? Optional.empty() : Optional.of(args.itemAt(0));
  }
}

File: cli/src/me/strangepan/tasks/cli/feature/add/AddParser.kt
package me.strangepan.tasks.cli.feature.add

import java.util.Optional
import omnia.data.cache.Memoized
import omnia.data.structure.List
import omnia.data.structure.immutable.ImmutableList
import org.apache.commons.cli.CommandLine
import me.strangepan.tasks.cli.parser.CommandParser
import me.strangepan.tasks.cli.parser.ParseResult
import me.strangepan.tasks.cli.parser.Parser
import me.strangepan.tasks.cli.parser.ParserException
import me.strangepan.tasks.cli.parser.ParserUtil
import me.strangepan.tasks.engine.model.Task

/** Command line argument parser for the Add command.  */
class AddParser(
    private val taskParser: Memoized<out Parser<out List<out ParseResult<out Task>>>>) : CommandParser<AddArguments> {
  override fun parse(commandLine: CommandLine): AddArguments {
    /*
     * 1st param must be description
     * 2+ params not supported
     * optional blocking me.strangepan.tasks.engine.tasks
     * optional blocked me.strangepan.tasks.engine.tasks
     */
    val argsList: List<String> = ImmutableList.copyOf(commandLine.argList)
    val taskDescription = extractTaskDescriptionFrom(argsList)
        .orElseThrow { ParserException("Task description not defined") }
    ParserUtil.assertNoExtraArgs(commandLine, AddCommand.COMMAND_PARAMETERS.value())
    val afterTasks = taskParser.value().parse(ParserUtil.getOptionValues(commandLine, AddCommand.AFTER_OPTION.value()))
    val beforeTasks = taskParser.value().parse(ParserUtil.getOptionValues(commandLine, AddCommand.BEFORE_OPTION.value()))

    // Initial validation combines before and after into a nice aggregate message
    ParserUtil.extractSuccessfulResultsOrThrow(
        ImmutableList.builder<ParseResult<*>>()
            .addAll(afterTasks)
            .addAll(beforeTasks)
            .build())
    return AddArguments(
        taskDescription,
        ParserUtil.extractSuccessfulResultsOrThrow(afterTasks),
        ParserUtil.extractSuccessfulResultsOrThrow(beforeTasks))
  }

  companion object {
    private fun extractTaskDescriptionFrom(args: List<String>): Optional<String> {
      return if (args.count() < 1) Optional.empty() else Optional.of(args.itemAt(0))
    }
  }
}